From 39b3c4a1975d91c698b076e98bdc6e0a92a83b46 Mon Sep 17 00:00:00 2001
From: AUR 7zip-natspec package <https://aur.archlinux.org/packages/7zip-natspec>
Date: Thu, 1 Jan 2026 00:00:00 +0000
Subject: [PATCH] Decode legacy ZIP entry names via libnatspec

Vendored from the AUR 7zip-natspec patch, pinned at repository commit
39b3c4a1975d91c698b076e98bdc6e0a92a83b46. Line endings normalized to LF:
7-Zip sources are distributed with LF while the AUR patch uses CRLF, so
the pristine patch fails to apply with "(different line endings)".
---
--- a/CPP/7zip/7zip_gcc.mak
+++ b/CPP/7zip/7zip_gcc.mak
@@ -162,7 +162,7 @@
 # LOCAL_LIBS=-lpthread
 # LOCAL_LIBS_DLL=$(LOCAL_LIBS) -ldl
 LIB2 = -lpthread
-LIB2 = -lpthread -ldl
+LIB2 = -lpthread -ldl -lnatspec
 
 
 endif
--- a/CPP/7zip/Archive/Zip/ZipItem.cpp
+++ b/CPP/7zip/Archive/Zip/ZipItem.cpp
@@ -1,5 +1,8 @@
 // Archive/ZipItem.cpp
 
+#include <cstdlib>
+#include <natspec.h>
+
 #include "StdAfx.h"
 
 #include "../../../../C/CpuArch.h"
@@ -451,6 +454,14 @@
     #endif
   }
   
+  if (!isUtf8) {
+    const char *oem_charset = natspec_get_charset_by_locale(NATSPEC_DOSCS, "");
+    char *p = natspec_convert(s.Ptr(), NULL, oem_charset, 0);
+    AString utf{p};
+    free(p);
+    if (ConvertUTF8ToUnicode(utf, res))
+      return;
+  }
   
   if (isUtf8)
   {
